subprojects/docs/src/docs/userguide/migrating_from_ant.adoc

```diff
@@ -0,0 +1,328 @@
+// Copyright 2018 the original author or authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+[[migrating_from_ant]]
+= Migrating Builds From Apache Ant
+
+https://ant.apache.org/[Apache Ant] is a build tool with a long history in the Java world and it is still widely used. While flexible, it lacks conventions and many of the powerful features that Gradle can provide. So migrating to Gradle is often worth the effort.
+
+The biggest problem with migrating from Ant to Gradle is that there is no such thing as a standard Ant build! That makes it difficult to provide specific instructions. Fortunately, Gradle has some great integration features with Ant that can make the process smoother. And even migrating from https://ant.apache.org/ivy/[Ivy]-based dependency management isn't particularly hard because Gradle has a similar model based on <<dependency_management_terminology#sub:terminology_configuration,dependency configurations>> and it works with Ivy-compatible repositories.
+
+We will start by outlining the kinds of approaches you can take to migrating from Ant and offer some general guidelines.
+
+== General guidelines
+
+When you undertake to migrate a build from Ant to Gradle, you should keep in mind the nature of both what you already have and where you would like to end up. Do you want a Gradle build that mirrors the structure of the existing Ant build? Or do you want to move to something that is more idiomatic to Gradle? This is an important decision that will impact the difficulty of the migration and the benefits you accrue from switching to Gradle.
+
+To understand the implications, it may be useful to look at the two extremes:
+
+ 1. Using <<ant.html#sec:import_ant_build,`ant.importBuild()`>> to import the Ant build wholesale into Gradle.
++
+This approach is quick and simple and works for many Ant-based builds. Your targets become Gradle tasks that you can run, and the dependencies between targets are retained.
++
+The downside is that you lose the advantages of Gradle's conventions, many of its plugins, and many of its important features like incremental build, proper multi-project support, the build cache, and more.
+ 2. Creating the Gradle build from scratch.
```

Comment:
Surely there's a middle ground here where we can advise how to split one's Ant config files and convert incrementally?

Reply:
One possible approach is to break up the build itself into parts. Maybe they're migrating to microservices and can split a library out into a separate build.

Reply:
This bit [on the Ant page](https://docs.gradle.org/5.0-rc-1/userguide/ant.html) seems inspiring:

> You can use Gradle’s Ant integration as a path for migrating your build from Ant to Gradle. For example, you could start by importing your existing Ant build. Then you could move your dependency declarations from the Ant script to your build file. Finally, you could move your tasks across to your build file, or replace them with some of Gradle’s plugins. This process can be done in parts over time, and you can have a working Gradle build during the entire process.

Reply:
Furthermore, we can add an option to incrementally convert and link to `Migrating multi-project builds` subsection here.

Reply:
I've only done one or two migrations, but those definitely benefited from creating the Gradle build "from scratch". Large chunks of Ant build script can end up being a few lines in the Gradle build.

The danger of a phased migration within a single project is that you end up with a dog's breakfast of a Gradle build without many of the advantages. This is an opportunity to clean up the mess that Ant builds often descend into.

I'd be interested to hear what @breskeby thinks as I'm no authority on the subject. Perhaps it's important to distinguish between simple, single-project builds and monolithic builds that actually deal with multiple projects.